applications/nrf5340_audio/src/bluetooth/bt_content_control/bt_cont_ctrl.c

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "bt_cont_ctrl.h"
+
+#include <zephyr/zbus/zbus.h>
+
+#include "bt_media_control.h"
+#include "nrf5340_audio_common.h"
+#include "macros_common.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bt_cont_ctrl, CONFIG_BT_CONTENT_CONTROL_LOG_LEVEL);
+
+ZBUS_CHAN_DEFINE(cont_media_chan, struct content_control_msg, NULL, NULL, ZBUS_OBSERVERS_EMPTY,
+		 ZBUS_MSG_INIT(0));
+
+static void media_control_cb(bool play)
+{
+	int ret;
+	struct content_control_msg msg;
+
+	if (play) {
+		msg.event = MEDIA_PLAY;
+	} else {
+		msg.event = MEDIA_PAUSE;
+	}
+
+	ret = zbus_chan_pub(&cont_media_chan, &msg, K_NO_WAIT);
+	ERR_CHK_MSG(ret, "zbus publication failed");
+}
+
+int bt_cont_ctrl_play(void)
+{
+	int ret;
+	struct content_control_msg msg;
+
+	if (IS_ENABLED(CONFIG_BT_MCC) || IS_ENABLED(CONFIG_BT_MCS)) {
+		ret = bt_media_control_play();
+		if (ret) {
+			LOG_WRN("Failed to change streaming state");
+			return ret;
+		}
+		return 0;
+	}
+
+	msg.event = MEDIA_PLAY;
+
+	ret = zbus_chan_pub(&cont_media_chan, &msg, K_NO_WAIT);
+	ERR_CHK_MSG(ret, "zbus publication failed");
+
+	return 0;
+}
+
+int bt_cont_ctrl_pause(void)
+{
+	int ret;
+	struct content_control_msg msg;
+
+	if (IS_ENABLED(CONFIG_BT_MCC) || IS_ENABLED(CONFIG_BT_MCS)) {
+		ret = bt_media_control_pause();
+		if (ret) {
+			LOG_WRN("Failed to change streaming state");
+			return ret;
+		}
+		return 0;
+	}
+
+	msg.event = MEDIA_PAUSE;
+
+	ret = zbus_chan_pub(&cont_media_chan, &msg, K_NO_WAIT);
+	ERR_CHK_MSG(ret, "zbus publication failed");
+
+	return 0;
+}
+
+int bt_cont_ctrl_disconnected(struct bt_conn *conn)
+{
+	int ret;
+
+	if (IS_ENABLED(CONFIG_BT_MCC)) {
+		ret = bt_media_control_conn_disconnected(conn);
+		if (ret) {
+			LOG_ERR("bt_media_control_conn_disconnected failed with %d", ret);
+		}
+	}
+
+	return 0;
+}
+
+int bt_cont_ctrl_discover(struct bt_conn *conn)
+{
+	int ret;
+
+	if (IS_ENABLED(CONFIG_BT_MCC)) {
+		ret = bt_media_control_discover(conn);
+		if (ret) {
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+int bt_cont_ctrl_init(void)
+{
+	int ret;
+
+	if (IS_ENABLED(CONFIG_BT_MCS)) {
+		ret = bt_media_control_server_init(media_control_cb);
+		if (ret) {
+			LOG_ERR("MCS server init failed");
+			return ret;
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_BT_MCC)) {
+		ret = bt_media_control_client_init();
+		if (ret) {
+			LOG_ERR("MCS client init failed");
+			return ret;
+		}
+	}
```

Comment:
Print warning if both CONFIG_BT_MCS and CONFIG_BT_MCC are disabled

Reply:
No, we've already talked about this, there are more services coming and this is called regardless if you are BIS or CIS at this stage. If there was a warning here then all BIS gateways and headset would print a warning at boot stating that some service it might never support is disabled

Reply:
I don't like that we can return 0 now, even if nothing has been initialized. Doing bt_content_control_init() with none of the configs enabled should to me trigger some kind of warning (either a LOG or an error code return).
In the current streamctrl we can guard against e.g. BIS doing bt_content_control_init() by having an if(IS_ENABLED(CONFIG_TRANSPORT_CIS)).

The same concept is also relevant for other "customer-facing functions" like bt_content_control_conn_disconnected(), bt_content_control_discover(), bt_mgmt_pa_sync_delete() and bt_mgmt_conn_disconnect().

Reply:
But you still call bt_content_control_start/stop even if you're BIS, so then it doesn't make sense to have a warning for the init. I can see the point on the very specific functions like  bt_content_control_conn_disconnected() but not for the general ones. It also goes against the point where we shouldn't add more API later. In the future there might be relevant functions here for BIS. You could potentially connect to a remote or a watch that would control the content of the gateway etc.

bt_content_control_conn_disconnected(), bt_content_control_discover(), bt_mgmt_pa_sync_delete() and bt_mgmt_conn_disconnect() all require a parameter when they are called. This parameter would not be available if you are not in the correct configuration.